common/oauthTokenManager.go

```diff
@@ -300,38 +326,111 @@ func (uotm *UserOAuthTokenManager) UserLogin(tenantID, activeDirectoryEndpoint s
 	return nil
 }
 
-// getCachedTokenInfo get a fresh token from local disk cache.
-// If access token is expired, it will refresh the token.
-// If refresh token is expired, the method will fail and return failure reason.
-// Fresh token is persisted if access token or refresh token is changed.
+// getCachedTokenInfo retrieves a fresh token.
+// If a valid AuthenticationRecord exists, the token is obtained from the azidentity token cache.
+// If the access token has expired, it is refreshed.
+// If the refresh token has expired, the method fails and returns the reason for failure.
+// The fresh token is persisted if either the access token or refresh token has changed.
 func (uotm *UserOAuthTokenManager) getCachedTokenInfo(ctx context.Context) (*OAuthTokenInfo, error) {
-	hasToken, err := uotm.credCache.HasCachedToken()
+
+	// retrieveRecord gets an AuthenticationRecord from the local disk
+	record, err := retrieveRecord()
 	if err != nil {
-		return nil, fmt.Errorf("no cached token found, please log in with azcopy's login command, %v", err)
+		return nil, err
 	}
-	if !hasToken {
-		return nil, errors.New("no cached token found, please log in with azcopy's login command")
+
+	// If the AAuthenticationRecord is empty get a fresh token from our credential cache.
+	if record == (azidentity.AuthenticationRecord{}) {
+		hasToken, err := uotm.credCache.HasCachedToken()
+		if err != nil {
+			return nil, fmt.Errorf("no cached token found, please log in with azcopy's login command, %v", err)
+		}
+		if !hasToken {
+			return nil, errors.New("no cached token found, please log in with azcopy's login command")
+		}
+
+		tokenInfo, err := uotm.credCache.LoadToken()
+		if err != nil {
+			return nil, fmt.Errorf("get cached token failed, %v", err)
+		}
+
+		freshToken, err := tokenInfo.Refresh(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("get cached token failed to ensure token fresh, please log in with azcopy's login command again, %v", err)
+		}
+
+		// Update token cache, if token is updated.
+		if freshToken.Token != tokenInfo.AccessToken.Token {
+			tokenInfo.AccessToken = *freshToken
+			if err := uotm.credCache.SaveToken(*tokenInfo); err != nil {
+				return nil, err
+			}
+		}
+		return tokenInfo, nil
+	}
+
+	// Create oAuthTokenInfo from the existing AuthenticationRecord
+	oAuthTokenInfo := &OAuthTokenInfo{
+		Tenant:   record.TenantID,
+		ClientID: record.ClientID,
 	}
 
-	tokenInfo, err := uotm.credCache.LoadToken()
+	tc, err := oAuthTokenInfo.GetTokenCredential()
 	if err != nil {
-		return nil, fmt.Errorf("get cached token failed, %v", err)
+		return nil, err
 	}
+	scopes := []string{StorageScope}
 
-	freshToken, err := tokenInfo.Refresh(ctx)
+	token, err := tc.GetToken(context.TODO(), policy.TokenRequestOptions{
```

Comment:
no, Authentication-record is passed whenever we are creating credential, if you observe at line no 373 we are creating Oauthtokeninfo based on the  Authentication-record , and then passing it over the GetTokenCredential() method where in case of Device Code credential its passing in the auth-record for tokenCredential creation.